src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java

```diff
@@ -265,6 +267,59 @@ public boolean canRecipeBeAdded(final IToken<?> ignored)
         return Optional.empty();
     }
 
+    /**
+     * Has a chance to reduce the resource requirements for the recipe in this building
+     * @param recipe the recipe we're possibly improving
+     * @param count the number of items (chances) 
+     * @param citizen The citizen, as the primary skill can improve the chances
+     */
+    public void improveRecipe(IRecipeStorage recipe, int count, ICitizenData citizen)
+    {
+        final double baseChance = 0.0625;
+        final ResourceLocation reducableIngredients = new ResourceLocation("minecolonies", "reduceable_ingredient");
+        final ResourceLocation reducableProductExclusions = new ResourceLocation("minecolonies", "reduceable_product_excluded");
+        final List<ItemStorage> inputs = recipe.getCleanedInput().stream().sorted(Comparator.comparingInt(ItemStorage::getAmount).reversed()).collect(Collectors.toList());
+
+        final double actualChance = Math.min(5.0, (baseChance * count) + (baseChance * citizen.getCitizenSkillHandler().getLevel(getPrimarySkill())));
+        final double roll = citizen.getRandom().nextDouble() * 100;
+
+        Log.getLogger().info(this.getJobName() + ": for " + recipe.getPrimaryOutput().getItem() + " improvement roll is " + roll + " with an actual chance of: " + actualChance);
+        if(roll <= actualChance && !ItemTags.getCollection().getOrCreate(reducableProductExclusions).contains(recipe.getPrimaryOutput().getItem()))
+        {
+            ArrayList<ItemStack> newRecipe = new ArrayList<>();
+            boolean didReduction = false;
+            Log.getLogger().info(this.getJobName() + ": attempting to improve recipe for: " + recipe.getPrimaryOutput().getItem());
+            for(ItemStorage input : inputs)
```

Comment:
This seems to go through ALL ingredients to improve then. But it should pick one at random actually.

Reply:
It only improves one, and while it's not random, it will improve the one with the largest starting count.  It touches all of the ingredients to copy them to the 'new' recipe.

Reply:
kk